NameLayerMain/src/vg/civcraft/mc/namelayer/group/Group.java

```diff
@@ -395,23 +396,38 @@ public static boolean areLinked(Group supergroup, Group subgroup) {
 		if (supergroup == null || subgroup == null) {
 			return false;
 		}
-		
-		return supergroup.hasSubGroup(subgroup) 
-				&& (subgroup.hasSuperGroup() 
-				&& subgroup.supergroup.equals(supergroup));
+		List <String> names = new LinkedList<String>();
+		names.add(supergroup.getName());
+		//check upwards to prevent cycles
+		if (supergroup.hasSuperGroup()) {
+			names.addAll(getRecursiveNamesUpwards(supergroup));
+		}
+		if (names.contains(subgroup.getName())) {
+			return true;
+		}
+		return false;
+	}
+	
+	private static List <String> getRecursiveNamesUpwards(Group g) {
```

Comment:
Not really safe; future code might accidentally introduce cycles, and this _will_ hardlock the server if triggered in the presence of one ...

Good news is that the non-recursive check is easier. Sketch:

```
// base check (are they the same)
if (subgroup.getName().equals(supergroup.getName()) return true;

// walk up the supergroups, checking for the subgroup as we go
Group superG = supergroup;
while (superG.hasSuperGroup() ) {
    if (superG.getName().equals(subgroup.getName()) return true;
    superG = superG.getSuperGroup();
}

// done
```